File: leetcode/5.最长回文子串.cpp
/*
 * @lc app=leetcode.cn id=5 lang=cpp
 *
 * [5] 最长回文子串
 */
#include <iostream>
#include <utility>
#include <string>
#include <string.h>
#include <vector>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;

// @lc code=start
class Solution {
public:
    string longestPalindrome(string s) {
        int l = 0;
        int len = 0;
        for (int i = 0; i < s.size(); ++i) {
            int left = i - 1;
            int right = i + 1;
            while (left >= 0 && right < s.size() && s[left] == s[right]) {
                --left;
                ++right;
            }
            int nLen = (right - left - 1);
            if (nLen > len) {
                l = left + 1;
                len = nLen;
            }
        }
        for (int i = 1; i < s.size(); ++i) {
            if (s[i] == s[i - 1]) {
                int left = i - 2;
                int right = i + 1;
                while (left >= 0 && right < s.size() && s[left] == s[right]) {
                    --left;
                    ++right;
                }
                int nLen = (right - left - 1);
                if (nLen > len) {
                    l = left + 1;
                    len = nLen;
                }
            }
        }
        return s.substr(l, len);
    }

    string longestPalindrom1(string s) {
        vector<vector<bool>> dp(s.size(), vector<bool>(s.size(), true));
        int left = 0, sz = 1;
        for (int j = 0; j < s.size(); ++j) {
            for (int i = 0; i < j; ++i) {
                dp[i][j] = ((s[i] == s[j]) & dp[i + 1][j - 1]);
                if (dp[i][j] && j - i + 1 > sz) {
                    sz = j - i + 1;
                    left = i;
                }
            }
        }
        return s.substr(left, sz);
    }
};
// @lc code=end

int main() {
    cout << Solution{}.longestPalindrome("babad");
}

File: python/5.最长回文子串.py
#
# @lc app=leetcode.cn id=5 lang=python3
#
# [5] 最长回文子串
#

# @lc code=start
def len_palindrome(s: str, left, right):
    if left < 0 or right >= len(s):
        return 0
    if s[left] != s[right]:
        return 0
    ans = 1
    if left != right:
        ans += 1
    left -= 1
    right += 1
    while left >= 0 and right < len(s):
        if s[left] == s[right]:
            left -= 1
            right += 1
            ans += 2
        else:
            break
    return ans

class Solution:
    def longestPalindrome(self, s: str) -> str:
        ans = 0
        left = 0
        right = 0
        for i in range(len(s)):
            oi = len_palindrome(s, i, i)
            ti = len_palindrome(s, i, i + 1)
            if oi > ans:
                ans = oi
                left = i
                right = i
            if ti > ans:
                ans = ti
                left = i
                right = i + 1
        return s[right - ans // 2 :left + ans // 2 + 1]
# @lc code=end

